src/main/java/com/rsr5/gocd/riemann_notifier/GoNotificationPlugin.java

```diff
@@ -138,23 +144,68 @@ private GoPluginApiResponse handleNotificationsInterestedIn() {
         return renderJSON(SUCCESS_RESPONSE_CODE, response);
     }
 
-    private String service(JsonObject json) {
-        JsonObject pipelineObject, stageObject;
-        pipelineObject = (JsonObject) json.get("pipeline");
-        stageObject = (JsonObject) pipelineObject.get("stage");
+    private String safeGetAsString(JsonObject json, String property) {
+        JsonElement element = json.get(property);
+        if (element != null && element.isJsonPrimitive()) {
+            return element.getAsString();
+        }
+        return null;
+    }
+
+    final class StageData {
+        public String pipeline;
+        public String stage;
+        public String state;
+        public String counter;
+        public Instant start;
+        public Instant end;
+        public List<JobData> jobs;
+
+        final class JobData {
+            public String name;
+            public String result;
+            public Instant start;
+            public Instant end;
+
+            public JobData(JsonObject json) {
+                this.name = safeGetAsString(json, "name");
+                this.result = safeGetAsString(json, "result");
+                this.start = Instant.parse(safeGetAsString(json, "schedule-time"));
+                this.end = Instant.parse(safeGetAsString(json, "complete-time"));
+            }
 
-        String pipeline = pipelineObject.get("name").getAsString();
-        String stage = stageObject.get("name").getAsString();
+            public long getDurationInSeconds() {
+                return ChronoUnit.SECONDS.between(this.start, this.end);
+            }
+        }
 
-        return "gocd:" + pipeline + ":" + stage;
-    }
+        public StageData(JsonObject json) {
+            JsonObject pipelineObject = (JsonObject) json.get("pipeline");
+            JsonObject stageObject = (JsonObject) pipelineObject.get("stage");
+
+            this.pipeline = safeGetAsString(pipelineObject, "name");
+            this.counter = safeGetAsString(pipelineObject, "counter");
+            this.stage = safeGetAsString(stageObject, "name");
+            this.state = safeGetAsString(stageObject, "state");
+            this.start = Instant.parse(safeGetAsString(stageObject, "create-time"));
+            this.end = Instant.parse(safeGetAsString(stageObject, "last-transition-time"));
+            this.jobs = new ArrayList<JobData>();
```

Comment:
As a best practice, you could use `Collections.unmodifiableList` to make it read-only after the class is constructed.